### PR TITLE
Dynamically generate data-l10n-id for content from YAML (closes #1340).

### LINKED ISSUE #1340: Figure out L10n for YAML flat file content
Moving to flat files & static site generation, we have content in YAML that we want localized. What do?

Here's an idea:
- Take the filename and property names in YAML as L10N string ID prefixes
- Hash the string content, use that as L10N string ID suffix
- Generate an en-us .ftl file from the YAML content files using those L10N string IDs
- Use those L10N string IDs in HTML page content, both in client-side React and static site generation.

So, for example:
- `content-src/experiments/activity-stream.yaml` has a `title` property with content "Activity Stream"
- the L10N string ID for that could be `activity-stream-title-bc6494e4f9589713c63fc6671e19ecd2`
- generate `.ftl` using that string ID
- include that string ID in `data-l10n-id` in markup (not sure how)

The idea here is that we have L10N IDs that naturally change when content changes, so that localizers using Pontoon see when there are new strings that need translation. That is, say the title changes to "Activity Stream is Awesome" - the L10N ID changes to `activity-stream-title-4a490ec33f2887a662f7c31ab29edec7`

Better ideas are totally welcome!

### PR DESCRIPTION
**Pull request by chuckharmston**
As discussed, this has been updated to allow for per-field l10n ID suffixes. So if the title of Activity Stream is updated to be plural again, the YAML can be updated to this:

``` yaml
title: 'Activity Streams'
title_l10nid: 'plural'
```

And its `data-l10n-id` will be changed from `activitystreamTitle` to `activitystreamTitlePlural`, invalidating all the translations. This also works for nested properties, e.g.:

``` yaml
contributors:
  -
    title: 'Staff Engineer'
    title_l10nid: 'staff'
```

Note: when this is merged, we will need to go through Pontoon and translate all the YAML content in English. This will always have to be done for new experiments.

### CODE REVIEW

**Comment by chuckharmston** [comment]
The alternative I was trying to describe is to add a l10n-id field in the YAML for each field. If present, it would be used in the same way that `l10n_version` is used here, but only for that field. The problem is that that's another thing to send over XHR, another piece of state we need to hold, and another thing we need to take into account when looking at changes to the DOM. It would have a noticeable impact on performance.

``` yaml
description: My awesome new description.
description-l10nid: 2016-10-28
```

**Comment by dannycoates** [comment]
original discussion #1340

**Comment by dannycoates** [comment]
> The problem is that that's another thing to send over XHR, another piece of state we need to hold, and another thing we need to take into account when looking at changes to the DOM

Seem like that's the price of admission if we don't want to hash the content and be able to change individual fields without invalidating the whole thing.

> It would have a noticeable impact on performance

How much?

**Comment by chuckharmston** [comment]
Code and PR have been updated to reflect earlier discussions. This is now ready for review.

**Comment by dannycoates** [comment]
I see how this takes care of populating the `data-l10n-id` attributes, but how are we planning on handling the translation files? Is there going to be like `/locales/en-US/tab-center.ftl` or `/locales/en-US/experiments.ftl`? Also how will we populate the default one, generate it from yaml?

**Comment by chuckharmston** [comment]
As I understand it, Pontoon will allow anything with the data-l10n-id attribute to be localized, even if it doesn't have an existing in the .ftl.

**Comment by dannycoates** [comment]
> As I understand it, Pontoon will allow anything with the data-l10n-id attribute to be localized, even if it doesn't have an existing in the .ftl.

Sure, but how do we consume those translations on the site? We have to point to something, right? Like: `<link rel="localization" href="/static/locales/{locale}/app.ftl">` Will all the experiment translations go into `app.ftl`?

**Comment by lmorchard** [comment]
From @chuck:

> Note: when this is merged, we will need to go through Pontoon and translate all the YAML content in English. This will always have to be done for new experiments.

From @dannycoates:

> how are we planning on handling the translation files? Is there going to be like /locales/en-US/tab-center.ftl or /locales/en-US/experiments.ftl? Also how will we populate the default one, generate it from yaml?

FWIW, what I had in mind for #1340 was to auto-generate an en-us `experiments.ftl` from the YAML content files using the L10N string IDs - i.e. as a task in gulp. Manual translation in Pontoon would be redundant, because the original content is already localized to en-US.

**Comment by chuckharmston** [comment]
> FWIW, what I had in mind for #1340 was to auto-generate an en-us experiments.ftl from the YAML content files using the L10N string IDs - i.e. as a task in gulp. Manual translation in Pontoon would be redundant, because the original content is already localized to en-US. 

I don't think there's anything stopping us from doing that.

**Comment by lmorchard** [comment]
> Sure, but how do we consume those translations on the site? We have to point to something, right? Like: <link rel="localization" href="/static/locales/{locale}/app.ftl"> Will all the experiment translations go into app.ftl?

I think L20N is okay with multiple `<link rel="localization">` references, so I figured we could also have `<link rel="localization" href="/static/locales/{locale}/experiments.ftl">` alongside the existing `app.ftl`. How that interacts with Pontoon, though, I'm not sure. That is, would Pontoon commit changes to the appropriate FTL file (`app.ftl` or `experiments.ftl`)

**Comment by lmorchard** [comment]
> I don't think there's anything stopping us from doing that.

Does that mean you plan to do the gulp task as part of this PR? If not, we should keep #1340 open and plan the gulp task as follow-on work.

**Comment by chuckharmston** [comment]
> Does that mean you plan to do the gulp task as part of this PR? If not, we should keep #1340 open and plan the gulp task as follow-on work.

I don't have plans for that, no, but nothing here precludes it. I wouldn't want to invest time into anything related to gulp until we've figured out if we even want to keep it (#1463)

**Comment by lmorchard** [comment]
> I wouldn't want to invest time into anything related to gulp until we've figured out if we even want to keep it 

We're keeping it so far, and we need this l10n stuff before we have time to really move significantly on Webpack - i.e. this sprint or asap. And whatever we do for generating the ftl can probably be ported over to webpack, even if only as a separate script or reusable lib

**Comment by lmorchard** [comment]
Also FWIW - we do need _something_ for the en-US .ftl after merging this PR, because _all_ content for experiments gets replaced with l10n IDs when the strings are missing. I'd rather this were done with an automated task rather than have us copypasta all the YAML strings by hand

screenshot 2016-10-31 15 59 54

**Comment by chuckharmston** [comment]
Yep, covered that in comment 0:

> Note: when this is merged, we will need to go through Pontoon and translate all the YAML content in English. This will always have to be done for new experiments.

**Review by lmorchard — CHANGES_REQUESTED** [review]
TL;DR: We can't ship this as-is, because all the experiment strings are missing from the en-US locale and L20n just replaces content with l10n-ids. 

So I'd want to see an en-US .ftl file ship along with this PR, ideally generated from the YAML via script or build task. Copying & pasting the content into Pontoon is just a roundabout way to get to the same place, IMO (i.e. an en-US ftl file).

**Comment by lmorchard** [comment]
> Yep, covered that in comment 0:

Yes, I know. And I'm saying we need to fix that as part of this PR, or in this sprint / before next deploy to close out #1340.

File: content-src/experiments/activity-stream.yaml
Comment (chuckharmston) [inline content-src/experiments/activity-stream.yaml:46]: Deleted all these empty headlines, which seem to be an artifact from Django.

File: frontend/src/app/lib/utils.js
Comment (chuckharmston) [inline frontend/src/app/lib/utils.js:3]: In order to reuse this logic in the gulp tasks, I had to move these to their own module that doesn't require babelification. I imported them here for testing and other purposes.

File: frontend/test/app/containers/ExperimentPage-test.js
Comment (chuckharmston) [inline frontend/test/app/containers/ExperimentPage-test.js:45]: Renamed to `_l10nsuffix` per @clouserw request.

**Comment by chuckharmston** [comment]
Updated to include the FTL generation task. [Example output](https://gist.github.com/chuckharmston/9d78bf95e7082a6b8a7617ad2182dbcf).

This is ready for review again.

**Comment by dannycoates** [comment]
When reordering a list like contributors, it doesn't look obvious that I'd need to add a suffix to `title` when the text isn't changing.

How do we plan on communicating that changing the order of a list requires adding or changing an `_l10nsuffix` to translated strings within those objects? (just reviewing PRs?)

**Review by lmorchard — APPROVED** [review]
:+1: LGTM! Tests are thorough & pass, all content looks good, code is good. Looks like it needs a rebase, but good to go otherwise

File: frontend/src/index.html
Comment (lmorchard) [inline frontend/src/index.html]: Yeah, I think we realized today that this file doesn't even need to exist :/

**Comment by chuckharmston** [comment]
> When reordering a list like contributors, it doesn't look obvious that I'd need to add a suffix to title when the text isn't changing.

I'm not really sure. My original PR had the ability to replace the index with a slug of sorts (e.g. contributor name), but there wasn't a reliable way to do that and also generate keys in both the Gulp task and component.

**Comment by lmorchard** [comment]
Hmm, the reordering concern is interesting, though maybe more a doco issue? 

I had at one point considered using a hash of the original string as l10nID as default suffix, but that seemed possibly fraught with its own issues

**Comment by lmorchard** [comment]
Also maybe we just try to avoid reordering as much as possible. Say we want to add a contributor and the list was originally in alpha order - just tack the name onto the end and sort on the client side? (Of course then we have to track the original index on the client vs the new sort order, but bleh)

**Pull request merged**
